SFTtech/openage
# cython-based C++-/Python interface

- openage is now a shared library
 - moved `py/openage` -> `openage`
 - moved `py/openage/codecompliance` -> `buildsystem/codecompliance`
 - moved `cpp/` -> `libopenage/`
 - moved `doc/implementation/` -> `doc/code`
 - integrated Cython into buildsystem
 - C++ -> Py function calling
  - C++ API is specified in PXD annotations in `.h` files
  - cpp/pyinterface/pyfunc.h
   - Func<returntype, arg0type, arg1type, ...>
   - Callable objects
   - Designed to be filled with pointers to cdef functions
   - __don't use raw function pointers in the interface__
  - openage/cppinterface/setup.py
   - setup() Fills the PyFunc pointers
   - must be called before using libopenage functionality
 - Exception translation
  - cpp/pyinterface/exctranslate.h
   - translate_exc_cpp_to_py
   - translate_exc_py_to_cpp
  - py/cppinterface/exctranslate.py
   - describe_exception
   - raise_cpp_error
  - C++ functions that are declared 'except +' if they may throw
  - Python functions that are declared 'except *' if they may raise
  - GIL-safe
  - everything is translated back and forth
   - metadata
   - message
   - backtrace
 - Documentation

C++:

 - Overhauled exception code
  - now lives in `error/error.h`
  - exceptions automatically store their cause exception
  - exceptions can collect backtraces
  - see demo in `cpp/error/demo.cpp`
  - Hierarchies may now be used
  - `doc/code/exceptions`
 - Moved util/error.h to error/error
 - Added error/handlers.cpp, which provides a SIGSEGV and terminate handler
 - refactored LZXD code
  - removed many C-isms and micro-optimizations
  - simplified codeflow, split methods into smaller ones
  - file split into multiple smaller ones (`cpp/util/compress`)
  - `decompress` replaced by `decompress_frame`, which decompresses just
    32768 bytes at a time instead of everything at once.
  - exposed to Python using the PyInterface
 - Simplified C++ logging
  - ExternalStringStream: like stringstream, but appends to a custom buffer
 - Modernized C++ testing
  - now uses TestError (e.g. via the `TESTFAIL` macro) instead of return values.
 - C++ exceptions can collect backtraces
 - `SIGSEGV` and `std::terminate` handlers
 - Enum class that allows adding new members at runtime, and associated data.
  - used by `log::lvl`
 - `util/init.h`: `OnInit` and `OnDeInit`: to call functions on dynamic initialization
 - fixed a bug with Timer

Python:

 - Bumped Python version requirement to 3.4
 - Massive clean-up
  - Unified all utility modules into the openage.util package
  - Removed some code that got obsoleted in the process
  - Made stuff more pylint-compliant
  - Replaced home-brewn Enum and NamedObject implementations by enum.Enum
 - Implemented some utilities
  - openage/util/fslike.py
   - FileSystemLikeObject abstract base class
   - Directory FileSystemLikeObject
  - openage/util/filelike.py
   - FileLikeObject abstract base class
   - StreamSeekBuffer FileLikeObject
   - StreamFragment FileLikeObject
   - FIFO FileLikeObject
  - FSLikeObject used by convert and codegen
 - Overhauled testing system
  - openage.testing.testing.assert_result
  - openage.testing.testing.assert_raises
  - openage.testing.testing.TestError
  - now done without CTest
  - C++ and Python tests and demos all called via Python
 - Integrated cabextract/cab.py with the rest of the project
  - now a FileSystemLikeObject
  - takes a FileLikeObject as an argument
  - added tests
 - PEFile has been refactored
  - resource strings are now resolved
  - the code uses NamedStructs
 - The Python logger now forwards to the C++ logger if the interface has been
   initialized
  - Central management of current loglevel
   - `--verbose`, `--quiet`
 - Program has one single entry point: `openage.__main__.main()`
  - run with `python3 -m openage`
  - alternatively: `./run`
  - subcommands: `game`, `test`, `codegen`
 - Media file conversion is now done when launching the game.

Buildsystem:

 - added valgrind-python.supp, a valgrind suppression file.
 - added basic whitespace compliance checking for textfiles
 - added pylint code compliance checking
  - Many files are not yet pylint-compliant; those have a header
    '# TODO pylint: disable=C,R,whatever categories'
 - added selective code compliance checking (closes #229)
 - re-worked Makefile targets

closes #217 

See the various doc files, including [pyinterface](https://github.com/mic-e/openage/blob/pyinterface/doc/code/pyinterface.md), [testing](https://github.com/mic-e/openage/blob/pyinterface/doc/code/testing.md)

Some things are still remaining to do, but they in no way prevent merging:

 - write interface for `FSLikeObject`
  - use that interface for all asset handling (in C++)
  - add `inotify` support to `FSLikeObject`
 - support for other sources for `openage.convert` (setup disk/.iso #82)

Note: If you wish to start work that derives from this before it's been merged, which you should, if you're about to start any new branch, base it on [this commit](https://github.com/mic-e/openage/commit/483c482649fa026cee8fe4024237f34f23255ab3). That commit is guaranteed to end up as-is in `openage/master`.
